### ISSUE TITLE
Plotly export options

### ISSUE THREAD

**Issue by TiemenSch** [imatlab/imatlab#6]
Dear Anntzer,

Thanks for this awesome implementation of the MATLAB kernel for Jupyter! While using this, I've come across the need for specifying the options that the fig2plotly function accepts. More specificly, I would like to be able to switch between interactive and static plots, but still inline! If I understand this correctly, one should normally be able to call fig2plotly with a specific extension in mind (e.g. PNG, SVG). 

Would it in any way be possible to change this (perhaps for the whole notebook at the time)?

Static images have the benefit of rendering everywhere, and will probably be succesfully exported from Jupyter to it's supported formats inline.

Best regards,
Tiemen

**Comment by anntzer** [comment]
Ultimately, all imatlab does is calling
```
                        fig2plotly(h, ...
                            'filename', builtin('sprintf', '%i', i), ...
                            'offline', true, ...
                            'open', false)
```
on each figure handle `h` (where `i` is an index variable), generating files `0, 1, ...` in a temporary folder, the displays them as HTML.

You should be able to override the `fig2plotly` function with a new function so that it instead generates e.g. a png file and a tiny html file that points to it.  Does that sound reasonable?  This seems preferable instead of exposing a bunch of individual knobs.

(If we go that route, I may change the function to e.g. `imatlab_plot_export` with a default setting to call `fig2plotly`.)

**Comment by TiemenSch** [comment]
My HTML knowledge is quite limited in that respect actually. I wouldn't mind changing the setting by hand hard-coded in the imatlab/plotly files if needed, but I would need some help where to look.  

I guess my ideal behaviour would be some (Matlab) command in the first cell like:
``imatlab('static')``
with possible other keywords being the filetypes or 'plotly' for full interactive figures. 

Another option would be magics, but I think the current 'magic-free' setup provides a very consistent experience. Just Matlab.

**Comment by anntzer** [comment]
I have some idea of how to do it, but no guarantees as to the timeline.

**Comment by anntzer** [comment]
See https://github.com/anntzer/imatlab#environment-variables, specifically `IMATLAB_EXPORT_FIG`.
attn @imatlab-helper: plotly export is no longer automatic, see linked.

**Comment by imatlab-helper** [comment]
Thanks for the notification @anntzer . I think the environment variable solution is great. This is a very reasonable way to support any time of matlab figure export mechanism, particularly since environment variables can be set on the fly.

A question on the defaults. Is there any reason we wouldn't want to set inline plotly offline mode as the default in the case the environment variable _isn't_ set? I imagine it's a pretty decent bet that most of the users that use imatlab with jupyter notebooks just want inline plotly figures. A simple option could be to check to see if imatlab is launched as a notebook, and if the environment variable isn't set, then set it to support plotly. Thoughts?

**Comment by anntzer** [comment]
I believe the current behavior leaves maximum flexibility to the user (for example, I'd probably default to not using any exporter at all).

**Comment by imatlab-helper** [comment]
Fair enough. Let me see if I can create some basic matlab functions that folks can add to their path that will help with this.

**Comment by anntzer** [comment]
Should be as simple as
```
function use_plotly
    setenv('IMATLAB_EXPORT_FIG', '...');
end
```

One of the reasons for not implementing these myself is that I have been very careful not to let the kernel "leak" anything function into either the MATLAB path, or variable into the MATLAB workspace (hence the slighlty horrible call syntax at https://github.com/anntzer/imatlab/blob/master/imatlab/_kernel.py#L224, which uses (effectively) `feval(@(t)t(i), get(0, 'children'))` to get the ith children without storing `get(0, 'children')` in an intermediate variable).

**Comment by imatlab-helper** [comment]
Understood, it makes sense. What if we provided a new folder in this repository that had those two functions that users could add/copy over to their matlab path?

**Comment by anntzer** [comment]
Sounds reasonable.  I think I would favor `imatlab_set_exporter('fig2plotly')` / `imatlab_set_exporter('print', ...)` with `'...'` passed as varargins to `print` (so it would typically start with e.g. `'-dpng', ...`)

**Comment by imatlab-helper** [comment]
Easy enough, I can put that together when I get a chance.

**Comment by anntzer** [comment]
Turns out my previous version would not work when exporting multiple figures (because they would disappear from `get(0, 'children')` upon closure).  I think the new version should work better, please confirm.

**Comment by imatlab-helper** [comment]
I took a stab at this, but it's not done. Sadly don't have the time to do further debugging right now. It's probably pretty close. It's patch2 in the forked branch on my account if anyone wants to pick up where I left it.

**Comment by imatlab-helper** [comment]
A couple updates on this env variable use.

1) The one for plotly works in a jupyter notebook browser window. However, now once a plotly figure shows up, the notebook won't save properly. I get this ```Notebook validation failed``` error on save:
```
The save operation succeeded, but the notebook does not appear to be valid. The validation error was:
Notebook Validation failed: None is not of type 'object':
null
```

2) The setenv line for png is missing a semicolon at the end of the print line (otherwise returns an invalid matlab expression).

```
setenv('IMATLAB_EXPORT_FIG', func2str(@(h) eval([ ...
    'print(h, [tempname(''.''), ''.png''], ''-dpng'');', ...
    'close(h);'])));
```

**Comment by anntzer** [comment]
First issue is now fixed.

**Comment by imatlab-helper** [comment]
I think we can consider this fixed.

**Comment by anntzer** [comment]
Released v0.2.
